src/backend/distributed/commands/common.c

```diff
@@ -0,0 +1,274 @@
+/*-------------------------------------------------------------------------
+ *
+ * common.c
+ *
+ *    Most of the object propagation code consists of mostly the same
+ *    operations, varying slightly in parameters passed around. This
+ *    file contains most of the reusable logic in object propagation.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "catalog/objectaddress.h"
+#include "nodes/parsenodes.h"
+#include "tcop/utility.h"
+
+#include "distributed/commands.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/deparser.h"
+#include "distributed/listutils.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/metadata/dependency.h"
+#include "distributed/metadata/distobject.h"
+#include "distributed/multi_executor.h"
+#include "distributed/worker_transaction.h"
+
+
+/*
+ * PostprocessCreateDistributedObjectFromCatalogStmt is a common function that can be used
+ * for most objects during their creation phase. After the creation has happened locally
+ * this function creates idempotent statements to recreate the object addressed by the
+ * ObjectAddress of resolved from the creation statement.
+ *
+ * Since object already need to be able to create idempotent creation sql to support
+ * scaleout operations we can reuse this logic during the initial creation of the objects
+ * to reduce the complexity of implementation of new DDL commands.
+ */
+List *
+PostprocessCreateDistributedObjectFromCatalogStmt(Node *stmt, const char *queryString)
+{
+	const DistributeObjectOps *ops = GetDistributeObjectOps(stmt);
+	Assert(ops != NULL);
+
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	/* check creation against multi-statement transaction policy */
+	if (!ShouldPropagateCreateInCoordinatedTransction())
+	{
+		return NIL;
+	}
+
+	if (ops->featureFlag && *ops->featureFlag == false)
+	{
+		/* not propagating when a configured feature flag is turned off by the user */
+		return NIL;
+	}
+
+	ObjectAddress address = GetObjectAddressFromParseTree(stmt, false);
+
+	EnsureCoordinator();
+	EnsureSequentialMode(ops->objectType);
+
+	/* If the object has any unsupported dependency warn, and only create locally */
+	DeferredErrorMessage *depError = DeferErrorIfHasUnsupportedDependency(&address);
```

Comment:
Can we use `ErrorOrWarnIfObjectHasUnsupportedDependency` instead to 

i) Not to have separate definitions if the object is already distributed but just started to have unsupported dependency (I'm not sure about all the cases that can cause this but any object supporting `CREATE OR REPLACE` syntax may cause)
ii) Raise messages with different levels according to use case?

I was planning to do that with a separate PR but I've realized that this PR updates the object specific parts I'm planning to update.

Reply:
I can add that with subsequent PR as well.

Reply:
As we've just discussed at the sync, I'll make that change with a subsequent PR.